### ISSUE TITLE
Mac: 'git status' can miss newly created files whose file descriptor has not yet been closed

### ISSUE THREAD

**Issue by wilbaker** [Microsoft/VFSForGit#721]
Found this while running tests with `osacompile`.

It appears that the native library can receive `HandleNewFileInRootNotification` callbacks *after* git status runs for files that were created before `git status` was run:

**Commands**

```
~/GVFSTest/src/libraries>rm TestLib.scpt; osacompile -o TestLib.scpt -x -d TestLib.applescript
~/GVFSTest/src/libraries>git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
~/GVFSTest/src/libraries>git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   TestLib.scpt

no changes added to commit (use "git add" and/or "git commit -a")
```

**gvfs.mount logging**

```
PrjFSLib.HandleFileNotification: libraries/TestLib.scpt notificationType: PrjFS_NotificationType_PreDelete isDirectory: 0
[2019-01-28 15:35:34 -08:00] TryAcquireLockExternal {"LockRequest":"git status (57730)","IsElevated":false,"Result":"Accepted"}
HandleNewFileInRootNotification: .git/index.lock notificationType: PrjFS_NotificationType_NewFileCreated isDirectory: 0
PrjFSLib.HandleFileNotification: .git/index.lock notificationType: PrjFS_NotificationType_NewFileCreated isDirectory: 0
PrjFSLib.HandleFileNotification: .git/index.lock notificationType: PrjFS_NotificationType_PreDelete isDirectory: 0
[2019-01-28 15:35:34 -08:00] ReleaseExternalLockRequested
[2019-01-28 15:35:34 -08:00] ReleaseLockHeldByExternalProcess {"CurrentLockHolder":"git status (57730)","IsElevated":false,"EnlistmentId":"fd36b952a2ea415bb73249888c4ba054","Result":"Released","DurationMS":152,"LockHeldExternallyMS":152,"ParseGitIndexMS":0,"UpdatePlaceholdersMS":0,"BlobsDownloaded":0,"BlobDownloadTimeMS":0,"CommitsAndTreesDownloaded":0,"CommitsAndTreesDownloadTimeMS":0,"SizeQueries":0,"SizeQueryTimeMS":0}
HandleNewFileInRootNotification: libraries/TestLib.scpt notificationType: PrjFS_NotificationType_NewFileCreated isDirectory: 0
PrjFSLib.HandleFileNotification: libraries/TestLib.scpt notificationType: PrjFS_NotificationType_NewFileCreated isDirectory: 0
```

**Possible approach to fix the issue**

1. Notify VFS4G when `KAUTH_VNODE_ADD_FILE` is received for a directory
2. VFS4G adds the directory to a set of directories that must be scanned before providing git with its list of modified files.
3. In the virtualization hook, scan the set of directories (built in steps 1 and 2) for files that are not in modifiedPaths and add them to modifiedPaths.

(There might be some more nitty-gritty details to figure out with this approach).

**A second possible explanation**

It's possible there could be a bug in how the Mac callbacks interact with the feature @kewillford added to VFS4G that prevents temporary files from being included in modified paths.  This should be investigated first before digging into a more complicated solution (as described above).

**Comment by wilbaker** [comment]
@jrbriggs I have put this in "macOS Next" for now, but I think it's at the top of that list.  Please feel free to change this to the EA bucket if you think that's a more appropriate spot.

**Comment by jrbriggs** [comment]
@wilbaker I agree this should be in EA. Anything that would impact the trustworthiness of version control will immediately make the bar.

**Comment by wilbaker** [comment]
@kewillford I'm going to split the temporary file issue into its own work item.  I have confirmed the race issue described here can occur for totally new files.

**Comment by wilbaker** [comment]
> I'm going to split the temporary file issue into its own work item

Filed #725 for this.

**Comment by jeschu1** [comment]
Unfortunately this isn't related to the issue @kewillford saw.  Followed in the debugger and we're definitely getting the AddFile event until after status has run.  

I'm doing some digging on possible solutions.

**Comment by wilbaker** [comment]
Fixed by #734
